compendium/DeclarativeServices/src/ComponentContextImpl.cpp

```diff
@@ -103,6 +104,146 @@ ComponentContextImpl::GetProperties() const
   return configManagerPtr->GetProperties();
 }
 
+/**
+ * The ServiceInterfacePointerLookupInfo struct is used to pass back
+ * the shared_ptr from GetInterfacePointer in addition to a boolean
+ * which is used to determine whether or not the pointer was retrieved
+ * from the map or if the map is null.
+ * 
+ * wasFoundInMapOrMapEmpty:
+ *   - true, if the service pointer was in the map
+ *       or if the map if null (service can be either
+ *       valid or a nullptr)
+*    - false if the service pointer was not found
+ *       in the map (implies service is nullptr)
+ */
+struct ServiceInterfacePointerLookupInfo
+{
+  std::shared_ptr<void> service;
+  bool wasFoundInMapOrMapEmpty;
+
+  bool IsValid() const
+  {
+    return static_cast<bool>(service) == wasFoundInMapOrMapEmpty;
+  }
+};
+
+/**
+ * Returns a ServiceInterfacePointerLookup struct containing the service interface
+ * pointer from the InterfaceMapConstPtr given the interfaceid as the key.
+ * 
+ * This function has a similar implementation to ExtractService except it returns
+ * an instance of the ServiceInterfacePointerLookupInfo struct to do error handling
+ * if necessary.
+ * 
+ * - map, a InterfaceMap instance.
+ * - interfaceId, The interface id string.
+ * - return value: ServiceInterfacePointerLookupInfo
+ */
+static ServiceInterfacePointerLookupInfo GetInterfacePointer(
```

Comment:
Since it is marked static at the namespace scope, it gives the function "internal linkage" which means that it is only accessible within the translation unit (barring use from any other translation units). This was included since we wanted to ensure that users of cppmicroservices weren't able to call this function.

Reply:
I see. my original concern was with public API users, not cppmicroservices maintainers accessing the function. This allows you to make this a private member function instead of static. I think making the function non-static is a better choice.

Reply:
I agree with your decision on making it non-static, however I do not think it should be a private function member in the ComponentContextImpl class. It is only used as an implementation detail when using LocateService/Services and has a very niche use case. I'm going to remove the static but not add it in the ComponentContextImpl.hpp file.